### ISSUE TITLE
JSON-RPC authentication on CCU

### ISSUE THREAD

**Issue by danielperna84** [danielperna84/pyhomematic#36]
Currently fetching system-variables (or doing anything else via JSON-RPC) adds a request-overhead by authenticating and logging out for each request.
HA polls those variables every 60 seconds, so for 50 variables, the CCU has to handle the 50 requests for the data + the ones for authentication. This is very inefficient and may lead to stability-issues on the CCU itself.

**Comment by klada** [comment]
Is this issue still valid? Because when looking at the code I can see that in `getAllSystemVariables` you are actually using `SysVar.getAll()`, which fetches all variables in one request (+2 for login/logoff). I think this actually looks good :)

If you want to get rid of the 2 requests for authentication, you could also use the optional XML-API addon (`/addons/xmlapi/sysvarlist.cgi`). Since you already support that addon for resolving names this would not be an entirely new dependency.

This actually has 2 benefits:

- 1 HTTP request for getting all variables, no authentication
- You could also set string variables through the XML-API, which JSON-RPC does not allow

So supporting the XML-API as an alternative way for variable operations would actually have multiple benefits.

**Comment by danielperna84** [comment]
> Is this issue still valid?

Yes, because ideally we wouldn't authenticate every time we fetch variables. It would be more elegant to only authenticate once and periodically refresh the access token (or whatever is used, it's been a while since I've done that).

The XML-API would only be a workaround. Also people would need to install the addon. The goal is to have this working without relying on the users having specific addons installed. So it should stay the way it is, and ideally with a solution that gets rid of those redundant authorizations. I just didn't come up with a way of doing that properly.

**Comment by klada** [comment]
OK, I see. You'd need to call `Session.renew(_session_id_ )` every **n** minutes then, I guess. But of course only if no other requests have been made. But yeah, this is not trivial with the current architecture of `pyhomematic`.

// BTW: Setting string variables is not possible through JSON-RPC. So I still think using XML-API would actually have benefits, but I agree with you that people should not be forced to install any addon on their CCU.

**Comment by danielperna84** [comment]
Back then we have decided to not do the strings because it is not natively supported. It's also possible with Homegear. But it would be inconsistent to provide a service that allows setting strings, but that doesn't work everywhere. So we decided to only do what all platforms have in common.
Plus: in the long run most people will stop using variables (and automation) in the CCU anyway. It would be messy to track some automations that are done on the CCU and others that are done in HASS. In the end people will only use their CCU as a gateway on the hardware level. Like a lot of people are also doing with ioBroker.
